text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,285 @@
+
+Consider the following trait:
+
+```rust
+trait IntoIntIterator {
+    fn into_int_iter(self) -> impl Iterator<Item = u32>;
+}
+```
+
+The semantics of this are analogous to introducing a new associated type within the surrounding trait;
+
+```rust
+trait IntoIntIterator { // desugared
+    type IntoIntIter: Iterator<Item = u32>;
+    fn into_int_iter(self) -> Self::IntoIntIter;
+}
+```
+
+(In general, this associated type may be generic; it would contain whatever generic parameters are captured per the generic capture rules given previously.)
+
+This associated type is introduced by the compiler and cannot be named by users.
+
+The impl for a trait like `IntoIntIterator` must also use `impl Trait` in return position:
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    fn into_int_iter(self) -> impl Iterator<Item = u32> {
+        self.into_iter()
+    }
+}
+```
+
+This is equivalent to specify the value of the associated type as an `impl Trait`:
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    type IntoIntIter = impl Iterator<Item = u32>
+    fn into_int_iter(self) -> Self::IntoIntIter {
+        self.into_iter()
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Equivalent desugaring for traits
+
+Each `-> impl Trait` notation appearing in a trait fn return type is desugared to an anonymous associated type; the name of this type is a fresh name that cannot be typed by Rust programmers. In this RFC, we will use the name `$` when illustrating desugarings and the like.
+
+As a simple example, consider the following (more complex examples follow):
+
+```rust
+trait NewIntoIterator {
+    type Item;
+    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
+}
+
+// becomes
+
+trait NewIntoIterator {
+    type Item;
+
+    type $: Iterator<Item = Self::Item>;
+
+    fn into_iter(self) -> <Self as NewIntoIterator>::$;
+}
+```
+
+## Equivalent desugaring for trait impls
+
+Each `-> impl Trait` notation appearing in a trait impl fn return type is desugared to the same anonymous associated type `$` defined in the trait along with a function that returns it. The value of this associated type `$` is an `impl Trait`. 
+
+```rust
+impl NewIntoIterator for Vec<u32> {
+    type Item = u32;
+
+    fn into_iter(self) -> impl Iterator<Item = Self::Item> {
+        self.into_iter()
+    }
+}
+
+// becomes
+
+impl NewIntoIterator for Vec<u32> {
+    type Item = u32;
+    
+    type $ = impl Iterator<Item = Self::Item>;
+
+    fn into_iter(self) -> <Self as NewIntoIterator>::$ {
+        self.into_iter()
+    }
+}
+```
+
+## Impl trait must be used in both trait and trait impls
```

Comment:
is it required that the `impl` be exactly the same as the `trait`?

```rust
impl NewIntoIterator for Vec<u32> {
    type Item = u32;
    fn into_iter(self) -> impl Iterator<Item = u32> + 'static {
//                                                  ~~~~~~~~~ ok or not?
        self.into_iter()
    }
}
```

Reply:
If you write the "desugered" version today, you can specify stricter requirements than the trait.
```rust
#![feature(type_alias_impl_trait)]

trait MyIntoIterator {
    type Item;
    type Iterator: Iterator<Item = Self::Item>;

    fn my_into_iter(self) -> Self::Iterator;
}

impl<T> MyIntoIterator for Vec<T> {
    type Item = T;
    type Iterator = impl DoubleEndedIterator<Item = T> + ExactSizeIterator;

    fn my_into_iter(self) -> Self::Iterator {
        IntoIterator::into_iter(self)
    }
}
```